### PR TITLE
Remove BoundedVec half-impls in xcm

### LINKED ISSUE #6572: XCM: Should not contain BoundedVec half-impls
There are at least two places in the base `xcm` crate which contain half-implementations of `BoundedVec`; `BoundedVec` and associated types should be removed from Substrate into its own crate for use by XCM crate in place of these half-way implementations.

### PR DESCRIPTION
**Pull request by KiChjang**
Resolves #6572.

This PR removes the half-impls of bounded vectors in the `xcm` crate and replaces them with `BoundedVec` in `bounded-collections`.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2324369

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2324459

**Review by bkchr — APPROVED** [review]

File: xcm/src/v3/mod.rs [hunk 1]
```diff
@@ -200,14 +201,20 @@ pub mod prelude {
 	}
 }
 
-#[derive(Clone, Eq, PartialEq, Encode, Decode, Debug, TypeInfo)]
+parameter_types! {
+	pub MaxNameLen: u32 = 48;
```
Comment (ggwpez) [inline xcm/src/v3/mod.rs:205]:
```suggestion
	pub MaxPalletNameLen: u32 = 48;
```

File: xcm/src/v3/mod.rs [hunk 2]
```diff
@@ -227,44 +232,29 @@ impl PalletInfo {
...
-		} else {
-			v.truncate(MAX_DISPATCH_ERROR_LEN);
-			MaybeErrorCode::TruncatedError(v)
+	fn from(v: Vec<u8>) -> Self {
+		match BoundedVec::try_from(v) {
+			Ok(error) => MaybeErrorCode::Error(error),
+			Err(mut error) => {
+				error.truncate(MaxDispatchErrorLen::get() as usize);
+				let error = BoundedVec::try_from(error).expect("length of vec was truncated; qed");
+				MaybeErrorCode::TruncatedError(error)
```
Comment (ggwpez) [inline xcm/src/v3/mod.rs:256]:
```suggestion
				MaybeErrorCode::TruncatedError(BoundedVec::truncate_from(error))
```

**Review by mrcnski — APPROVED** [review]

File: xcm/src/v3/mod.rs (hunk 2 above)
Comment (KiChjang) [reply xcm/src/v3/mod.rs:256]: Ah right, knew I was forgetting a function, thanks!

**Comment by KiChjang** [comment]
Must bump substrate after paritytech/substrate#13243 is merged.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2330742

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2352701

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2353127

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2353171

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2356555

File: xcm/pallet-xcm-benchmarks/src/generic/benchmarking.rs [hunk 3]
```diff
@@ -360,9 +360,9 @@ benchmarks! {
 
 	expect_transact_status {
 		let mut executor = new_executor::<T>(Default::default());
-		// 1024 is an overestimate but should be good enough until we have `max_encoded_len`.
-		// Eventually it should be replaced by `DispatchError::max_encoded_len()`.
-		let worst_error = || MaybeErrorCode::Error(vec![0; 1024]);
+		let worst_error = || -> MaybeErrorCode {
+			vec![0; MaxDispatchErrorLen::get() as usize].into()
```
Comment (gavofyork) [inline xcm/pallet-xcm-benchmarks/src/generic/benchmarking.rs:364]: ❤️

**Review by gavofyork — APPROVED** [review]

**Comment by KiChjang** [comment]
Yes, that is the case.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2368554

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2368620

**Comment by KiChjang** [comment]
bot merge

**Pull request merged**